packages/firestore/test/unit/local/web_storage_metadata_channel.test.ts

```diff
@@ -0,0 +1,271 @@
+
+    expect(actual.lastUpdateTime).to.be.a('number');
+    expect(actual.lastUpdateTime).to.be.greaterThan(
+      Date.now() - GRACE_INTERVAL_MS
+    );
+    expect(actual.lastUpdateTime).to.be.at.most(Date.now());
+
+    expect(actual.activeTargetIds).to.be.an('array');
+    expect(actual.activeTargetIds).to.be.deep.equal(activeTargetIds);
+
+    if (minMutationBatchId !== undefined) {
+      ++expectedKeyCount;
+      expect(actual.minMutationBatchId).to.be.equal(minMutationBatchId);
+    }
+
+    if (maxMutationBatchId !== undefined) {
+      ++expectedKeyCount;
+      expect(actual.maxMutationBatchId).to.be.equal(maxMutationBatchId);
+    }
+
+    expect(Object.keys(actual).length).to.be.equal(expectedKeyCount);
+  }
+
+  describe('persists mutation batches', () => {
+    beforeEach(() => {
+      return persistenceHelpers
+        .testLocalStorageNotificationChannel(ownerId, [], [])
+        .then(nc => {
+          notificationChannel = nc;
+        });
+    });
+
+    afterEach(() => {
+      notificationChannel.shutdown();
+    });
+
+    it('with empty batch', () => {
+      assertInstanceState([]);
+    });
+
+    it('with one batch', () => {
+      notificationChannel.addLocalPendingMutation(0);
+      assertInstanceState([], 0, 0);
+    });
+
+    it('with multiple batches', () => {
+      notificationChannel.addLocalPendingMutation(0);
+      notificationChannel.addLocalPendingMutation(1);
+      assertInstanceState([], 0, 1);
+
+      notificationChannel.addLocalPendingMutation(2);
+      notificationChannel.addLocalPendingMutation(3);
+      assertInstanceState([], 0, 3);
+
+      notificationChannel.removeLocalPendingMutation(0);
+      notificationChannel.removeLocalPendingMutation(2);
+      assertInstanceState([], 1, 3);
+    });
+  });
+
+  describe('persists query targets', () => {
+    beforeEach(() => {
+      return persistenceHelpers
+        .testLocalStorageNotificationChannel(ownerId, [], [])
+        .then(nc => {
+          notificationChannel = nc;
+        });
+    });
+
+    afterEach(() => {
+      notificationChannel.shutdown();
+    });
+
+    it('with empty targets', () => {
+      assertInstanceState([]);
+    });
+
+    it('with multiple targets', () => {
+      notificationChannel.addLocallyActiveQueryTarget(0);
+      assertInstanceState([0]);
+
+      notificationChannel.addLocallyActiveQueryTarget(1);
+      notificationChannel.addLocallyActiveQueryTarget(2);
+      assertInstanceState([0, 1, 2]);
+
+      notificationChannel.removeLocallyActiveQueryTarget(1);
+      assertInstanceState([0, 2]);
+    });
+  });
+
+  describe('combines instance data', () => {
+    let previousAddEventListener;
+    let storageCallback: (StorageEvent) => void;
+
+    beforeEach(() => {
+      previousAddEventListener = window.addEventListener;
+      window.addEventListener = (type, callback) => {
+        expect(type).to.be.equal('storage');
+        storageCallback = callback;
+      };
```

Comment:
Not a big fan of this, but since we won't need to port it, maybe it's okay...  Would it be terrible to actually go through LocalStorage and rely on the real events?  That seems less hacky and provide better validation anyway (that local storage events actually work the way we think they do).

Reply:
This doesn't work for us since LocalStorage doesn't fire for local updates. Added comment.